### PR TITLE
Add functionalities to communication

### PR DESCRIPTION
**Pull request by ghost**
For code reviewer
Requirements : 
- add Streams (getting input/output stream)
- add Reader (reading from stream)
- add Writer (writing to stream)
- + implementations
- + tests

### CODE REVIEW

File: common/src/main/java/com/blackship/battlesheep/common/communication/Streams.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.blackship.battlesheep.common.communication;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Optional;
+
+public interface Streams {
+    Optional<OutputStream> getOutput();
+    Optional<InputStream> getInput();
```
Comment (arturAR) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Streams.java:9]: interesting 👍

File: common/src/main/java/com/blackship/battlesheep/common/communication/Reader.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.blackship.battlesheep.common.communication;
+
+import java.io.IOException;
+
+public interface Reader {
```
Comment (arturAR) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Reader.java:5]: Interface need javadoc!

File: common/src/main/java/com/blackship/battlesheep/common/communication/Writer.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.blackship.battlesheep.common.communication;
+
+import java.io.IOException;
+
+public interface Writer {
```
Comment (arturAR) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Writer.java:5]: Interface need javadoc!

File: common/src/main/java/com/blackship/battlesheep/common/communication/network/NetworkReader.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.blackship.battlesheep.common.communication.Reader;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class NetworkReader implements Reader {
+
+    private final InputStream inputStream;
```
Comment (arturAR) [inline common/src/main/java/com/blackship/battlesheep/common/communication/network/NetworkReader.java]: why final?

File: common/src/main/java/com/blackship/battlesheep/common/communication/Writer.java (hunk 3 above)
Comment (Devas) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Writer.java:5]: :+1:

File: common/src/main/java/com/blackship/battlesheep/common/communication/Reader.java (hunk 2 above)
Comment (Devas) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Reader.java:5]: :+1:

File: common/src/main/java/com/blackship/battlesheep/common/communication/Streams.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.blackship.battlesheep.common.communication;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Optional;
+
+public interface Streams {
```
Comment (Devas) [inline common/src/main/java/com/blackship/battlesheep/common/communication/Streams.java:7]: :+1:

File: common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkStreamsTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import static org.testng.Assert.*;
+
+public class NetworkStreamsTest {
+
+    private Socket givenSocket;
+    private NetworkStreams givenNetworkStream;
+
+    @BeforeTest
+    public void setUpSocket() {
+        givenSocket = mock(Socket.class);
```
Comment (Devas) [inline common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkStreamsTest.java:21]: :+1:

File: common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkWriterTest.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    @Test
+    public void writeAndExpectNoExceptionTest() throws IOException {
+
+        //given
+        Writer givenNetworkWriter = new NetworkWriter(new ByteArrayOutputStream(1));
+
+        //when
+        givenNetworkWriter.write(new byte[1]);
+
+        //then - expect no exceptions
```
Comment (Devas) [inline common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkWriterTest.java:22]: Not finished

File: common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkReaderTest.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import java.io.IOException;
+
+import static org.testng.Assert.*;
+
+public class NetworkReaderTest {
+
+    @Test
+    public void readAndExpectGivenStringTest() throws IOException {
+
+        //given
```
Comment (Devas) [inline common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkReaderTest.java:16]: I don't use these comments in tests but if it helps you it's ok

File: common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkStreamsTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import org.testng.annotations.Test;
+
+import java.io.*;
+import java.net.Socket;
+import java.util.Optional;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.*;
+
```
Comment (Devas) [inline common/src/test/java/com/blackship/battlesheep/common/communication/network/NetworkStreamsTest.java:13]: You can also write @Test before whole class

**Review by Devas — APPROVED** [review]

**Review by arturAR — APPROVED** [review]

**Pull request merged**